# Pathway Ensemble Tool: PET
<img width="781" alt="image" src="https://user-images.githubusercontent.com/16437494/207137637-32dec909-145c-4a3a-9421-57f62189dfb2.png">
PET is an ensemble pathway analysis tool which combines the results from three other pathway analysis methods (i.e. GSEA, ORA, Enrichr) optimized using Benchmark. 

## Requirement
* Python >= 3.6
* Numpy
* Scipy
* GSEApy
* PyDESeq2
* statsmodels
* DESeq2 (R version)

## Installation
To install PET, please refer to the [Installation](https://github.com/hedgehug/PET/blob/main/Installation.md) guide.

* If all dependencies are already installed, you can directly follow the [Tutorial](https://github.com/hedgehug/PET/blob/main/run_PET_tutorial.ipynb) to run PET.
* PET uses GSEApy by default, but if other versions of GSEA are preferred, you can download the **command line (all platforms)** version from [here](http://www.gsea-msigdb.org/gsea/downloads.jsp).

## Required input files
* Pathways to be investigated (e.g. GO biological processes) as [.gmt files](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#Gene_Set_Database_Formats). You can find an example pathway file [here](https://github.com/hedgehug/PET/blob/main/example/c2.cp.kegg.v2023.1.Hs.symbols.gmt). 
* Expression data from samples to be investigated, as a tab-delimited text file. **Raw counts** are highly recommended as required by DESeq2. The first column of the file should contain the gene symbols and the of rest columns should contain the expression values ordered by conditions. An example expression file is provided [here](https://github.com/hedgehug/PET/tree/main/example/example_data.txt). 
* A design matrix file defining the conditions and number of samples as a tab-delimited text file to be used by DESeq2. The first column should specify the condition name and the second column should indicate the number of samples. **The order and the number of samples should match the expression file above.** An example contrast file can be found [here](https://github.com/hedgehug/PET/tree/main/example/example_contrast.txt). 

## Run PET
To run PET, simply follow the step-by-step process outlined in the [Tutorial](https://github.com/hedgehug/PET/blob/main/run_PET_tutorial.ipynb).

## Evaluate pathway analysis method using Benchmark
Benchmark is a new platform that use data generated by high-throughput sequencing technologies (e.g. RNA-seq) from [ENCODE](https://www.encodeproject.org/) to evaluate the performance of pathway analysis tools. 
The data used to construct Benchmark can be found [here](https://github.com/hedgehug/PET/tree/main/data).  

| Assay                                | Source                                     | Number of targets | Curated target pathways                                                                                                                                                                                                                                                                                                                                                                    |
|--------------------------------------|--------------------------------------------|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| shRNA knockdown followed by RNA-seq | HepG2, K562                                |197| [HepG2_up](https://github.com/hedgehug/PET/blob/main/data/ENCODE_HepG2_RNA_up.gmt), [HepG2_down](https://github.com/hedgehug/PET/blob/main/data/ENCODE_HepG2_RNA_down.gmt), [K562_up](https://github.com/hedgehug/PET/blob/main/data/ENCODE_K562_RNA_up.gmt), [K562_down](https://github.com/hedgehug/PET/blob/main/data/ENCODE_K562_RNA_down.gmt)                                         |
| eCLIP-seq                          | HepG2, K562                                |73| [HepG2_pvalue](https://github.com/hedgehug/PET/blob/main/data/ENCODE_HepG2_eCLIP_pval.gmt), [HepG2_sigval](https://github.com/hedgehug/PET/blob/main/data/ENCODE_HepG2_eCLIP_signal_value.gmt), [K562_pvalue](https://github.com/hedgehug/PET/blob/main/data/ENCODE_K562_eCLIP_pval.gmt), [K562_sigval](https://github.com/hedgehug/PET/blob/main/data/ENCODE_K562_eCLIP_signal_value.gmt) |                                                                                                                                                   |
| ChIP-seq                           | HepG2, K562, CH12.LX, MEL, GM12878, A549, HEK293 |374| [ChIP_qvalue](https://github.com/hedgehug/PET/blob/main/data/ENCODE_ChIP_seq_qvalue.gmt), [ChIP_score](https://github.com/hedgehug/PET/blob/main/data/ENCODE_ChIP_seq_peak_score.gmt)                                                                                                                                                                                                                                                                                   |

Notes: 
For RNA-seq based data, the pathway files contain top 200 up/down-regulated differentially expressed genes after shRNA knockdown with specific target.
For eCLIP-seq and ChIP-seq based data, the pathway files contain annotated genes near the top peaks, sorted based on peak signal value/p-value/q-value.

To evaluate a new method, It should be executed using the formatted expression files from one cell line, e.g. [K562 RNA-seq](https://github.com/hedgehug/PET/blob/main/data/ENCODE_K562_RNA_seq_raw_count.zip), and the above curated pathway file in another cell line, e.g. [HepG2 RNA-seq](https://github.com/hedgehug/PET/blob/main/data/ENCODE_HepG2_RNA_down.gmt). 
To compare the results against other methods (e.g. ORA, Enrichr, GSEA), please follow the [tutorial](https://github.com/hedgehug/PET/blob/main/evaluate_new_method.ipynb)

The template scripts for running some of the methods used in this study are provided [here](https://github.com/hedgehug/PET/tree/main/template_script).

## Updates
* PyDESeq2 and GSEApy are now supported by PET.
* The selection of required DEGs for ORA or Enrichr can now be customized by user.
* PET can now support passing a design matrix directly to DESeq2.  

## Copyright
Copyright (c) 2023 Purdue University All rights reserved.

Developed by:  Kazemian lab (Luopin Wang), College of Science, Purdue University, https://kazemianlab.com/

Permission to use, copy, modify, and distribute this software and associated documentation (the “Software”) for educational, research and non-profit purposes, in each case by a government agency, educational institution, or other non-profit organization, without fee, and without a written agreement is hereby granted, provided that the above copyright notice, this paragraph and the following three paragraphs appear in all copies.
 
Permission to make commercial use of this Software or permission for any use by a for-profit entity may be requested by contacting:
Purdue Research Foundation
Office of Technology Commercialization
The Convergence Center
101 Foundry Dr., Suite 2500
West Lafayette, IN 47906
ATTN: Business Development Manager
otcip@prf.org
 
DISCLAIMER. THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE SOFTWARE.
 
LIMITATION OF LIABILITY. TO THE MAXIMUM EXTENT PERMITTED BY LAW, IN NO EVENT WILL PURDUE OR PURDUE RESEARCH FOUNDATION BE LIABLE TO ANY USER OF THIS CODE FOR ANY INCIDENTAL, CONSEQUENTIAL, EXEMPLARY OR PUNITIVE DAMAGES OF ANY KIND, LOST GOODWILL, LOST PROFITS, LOST BUSINESS AND/OR ANY INDIRECT ECONOMIC DAMAGES WHATSOEVER, REGARDLESS OF WHETHER SUCH DAMAGES ARISE FROM CLAIMS BASED UPON CONTRACT, NEGLIGENCE, TORT (INCLUDING STRICT LIABILITY OR OTHER LEGAL THEORY), A BREACH OF ANY WARRANTY OR TERM OF THIS AGREEMENT, AND REGARDLESS OF WHETHER PURDUE OR PURDUE RESEARCH FOUNDATION WAS ADVISED OR HAD REASON TO KNOW OF THE POSSIBILITY OF INCURRING SUCH DAMAGES IN ADVANCE.