{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1dba7033",
   "metadata": {},
   "source": [
    "This notebook is a tutorial for running PET.\n",
    "From the benchmark, we identified the optimal setting for running PET, which consists of 3 steps:\n",
    "1. perform differential expression analysis with DESeq2\n",
    "2. perform pathway analysis with underlying methods\n",
    "3. combine the result using PET\n",
    "\n",
    "File required:\n",
    "1. Data matrix, raw read count strongly recommended.\n",
    "2. Pathway file, in gmt format.\n",
    "\n",
    "Example data:\\\n",
    "We provided an example read count matrix text file and gmt pathway file (KEGG) as an example.\n",
    "\n",
    "Notes:\\\n",
    "Though template script and command for running GSEA and DESeq2 are provided, users should feel free to use any differential expression analysis methods and GSEA mode as wanted, as long as the output format is consistent for correctly parsing the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e4c4e48a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from fisher_test import run_fisher_test\n",
    "from enrichr import run_enrichr\n",
    "from PET import run_PET\n",
    "from helper import *\n",
    "import os, sys\n",
    "import numpy as np\n",
    "import scipy.stats as st"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "56a0face",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# create result directory\n",
    "out_dir = 'example/'\n",
    "create_dir(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4628400f",
   "metadata": {},
   "source": [
    "Step 1. \n",
    "Perform DEseq2 analysis, the template DESeq2 script is provided in deseq2_template.R.\\\n",
    "Note: **Raw read count** is required for DESeq2, first column as gene name, rest column as group1+group2 expression in order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad8f7dbb",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DESeq2 script written to deseq_analysis.R\n"
     ]
    }
   ],
   "source": [
    "# helper function to format DEseq2 script, please specify the result file with .csv suffix\n",
    "format_deseq2_script(read_count_file_path=out_dir+'example_data.txt', sample_num=10, group1_id='poor_survival',\n",
    "                     group2_id='better_survival', group1_num=5, group2_num=5, \n",
    "                     result_file_path=out_dir+'deseq_result.csv', script_name='deseq_analysis.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ecbf785",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h         TCGA.AA.3850 TCGA.AA.3845 TCGA.AA.3852 TCGA.G4.6302 TCGA.AA.3492\n",
      "TSPAN6           2447         2053         1454         1822         2356\n",
      "TNMD                9            3            3            2            4\n",
      "DPM1              490          876          392         1012          317\n",
      "SCYL3             255          299          246          468          301\n",
      "C1orf112          164          221          147          187          396\n",
      "FGR               159          152          137          534          115\n",
      "         TCGA.G4.6314 TCGA.G4.6311 TCGA.A6.2672 TCGA.DM.A1HA TCGA.DM.A1HB\n",
      "TSPAN6           8201         5085          747         2519         2463\n",
      "TNMD               36            2            3            1            0\n",
      "DPM1             2264         2981          265         3555         2408\n",
      "SCYL3             707          658          527          818          776\n",
      "C1orf112          575          845          249          386          430\n",
      "FGR               323          459          340           71          164\n",
      "\u001b[?25h [1] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      " [5] poor_survival   better_survival better_survival better_survival\n",
      " [9] better_survival better_survival\n",
      "Levels: better_survival poor_survival\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hclass: DESeqDataSet \n",
      "dim: 53657 10 \n",
      "metadata(1): version\n",
      "assays(1): counts\n",
      "rownames(53657): TSPAN6 TNMD ... LINC01144 AC007389.5\n",
      "rowData names(0):\n",
      "colnames(10): TCGA.AA.3850 TCGA.AA.3845 ... TCGA.DM.A1HA TCGA.DM.A1HB\n",
      "colData names(1): condition\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "# run DESeq2 script\n",
    "! Rscript deseq_analysis.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca355d66",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# extract DESeq2 result, CSV file only, to .rnk file for GSEA, please do check the direction of DESeq2 result (log2FoldChange column)\n",
    "# if log2FoldChange>=0 means up-regulation, keep direction=1; otherwise, pass direction=-1\n",
    "\n",
    "generate_rank_file(deseq_result_file=out_dir+'deseq_result.csv', out_file=out_dir+'prerank.rnk', direction=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7fbae8d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using system JDK.\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "504      [INFO  ] - Parameters passing to GSEAPreranked.main:\n",
      "505      [INFO  ] - rnk\texample/prerank.rnk\n",
      "505      [INFO  ] - gmx\texample/c2.cp.kegg.v2023.1.Hs.symbols.gmt\n",
      "505      [INFO  ] - rpt_label\texample_test\n",
      "505      [INFO  ] - collapse\tNo_Collapse\n",
      "505      [INFO  ] - zip_report\tfalse\n",
      "505      [INFO  ] - gui\tfalse\n",
      "505      [INFO  ] - out\texample/\n",
      "505      [INFO  ] - mode\tMax_probe\n",
      "506      [INFO  ] - norm\tmeandiv\n",
      "506      [INFO  ] - nperm\t1000\n",
      "506      [INFO  ] - scoring_scheme\tweighted\n",
      "506      [INFO  ] - include_only_symbols\ttrue\n",
      "506      [INFO  ] - make_sets\ttrue\n",
      "506      [INFO  ] - plot_top_x\t5\n",
      "506      [INFO  ] - rnd_seed\ttimestamp\n",
      "506      [INFO  ] - create_svgs\tfalse\n",
      "506      [INFO  ] - set_max\t500\n",
      "506      [INFO  ] - set_min\t15\n",
      "586      [INFO  ] - Made Vdb dir JIT: /Users/luopin/GSEA_test/ENCODE_RNA-seq/PET/mar11\n",
      "603      [INFO  ] - Begun importing: RankedList from: example/prerank.rnk\n",
      "995      [INFO  ] - Your current version of GSEA is 4.1.0. A newer version (4.3.2) is available.\n",
      "To update, please download from http://www.gsea-msigdb.org/gsea/gsea/downloads.jsp\n",
      "(build date: Mon, 3 Oct 2022 16:56:43 -0400)\n",
      " Better support for our new Mouse MSigDB release\n",
      "995      [INFO  ] - Note: GenePattern users should update through GenePattern.\n",
      "1040     [INFO  ] - No ranked list collapsing was done .. using original as is\n",
      "to parse>example/c2.cp.kegg.v2023.1.Hs.symbols.gmt< got: [example/c2.cp.kegg.v2023.1.Hs.symbols.gmt]\n",
      "1041     [INFO  ] - Begun importing: GeneSetMatrix from: c2.cp.kegg.v2023.1.Hs.symbols.gmt\n",
      "1066     [INFO  ] - Got gsets: 186 now preprocessing them ... min: 15 max: 500\n",
      "1082     [INFO  ] - Done preproc for smaller than: 15\n",
      "1198     [INFO  ] - Timestamp used as the random seed: 1678566691396\n",
      "shuffleGeneSet for GeneSet 1/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 6/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 11/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 16/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 21/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 26/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 31/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 36/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 41/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 46/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 51/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 56/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 61/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 66/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 71/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 76/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 81/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 86/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 91/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 96/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 101/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 106/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 111/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 116/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 121/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 126/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 131/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 136/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 141/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 146/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 151/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 156/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 161/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 166/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 171/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 176/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "29875    [INFO  ] - Creating marker selection reports ...\n",
      "29959    [INFO  ] - Creating FDR reports ...\n",
      "31426    [INFO  ] - Done FDR reports for positive phenotype\n",
      "31972    [INFO  ] - Done FDR reports for negative phenotype\n",
      "32060    [INFO  ] - Creating global reports ...\n",
      "32060    [INFO  ] - Done all reports!!\n",
      "Time taken: 31 secs\n"
     ]
    }
   ],
   "source": [
    "# To run GSEA, please download GSEA command line tool from here: http://www.gsea-msigdb.org/gsea/downloads.jsp\n",
    "# please change the gsea-cli.sh pathway as needed\n",
    "! ~/GSEA_test/GSEA_cmd/gsea-cli.sh GSEAPreranked -gmx example/c2.cp.kegg.v2023.1.Hs.symbols.gmt -collapse No_Collapse -mode Max_probe -norm meandiv -nperm 1000 -rnk example/prerank.rnk  -scoring_scheme weighted -rpt_label example_test   -create_svgs false -include_only_symbols true -make_sets true -plot_top_x 5 -rnd_seed timestamp -set_max 500 -set_min 15 -zip_report false -out example/  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd260f84",
   "metadata": {},
   "source": [
    "Before running other methods, we shall prune the pathways, which will drop pathways with gene_num > max_num and gene_num < min_num and also remove any gene that's not present in the expression matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "82553708",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53658 genes in the expression matrix\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES removed due to gene number\n",
      "KEGG_NON_HOMOLOGOUS_END_JOINING removed due to gene number\n",
      "KEGG_CIRCADIAN_RHYTHM_MAMMAL removed due to gene number\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS removed due to gene number\n",
      "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM removed due to gene number\n",
      "KEGG_FOLATE_BIOSYNTHESIS removed due to gene number\n",
      "KEGG_LIMONENE_AND_PINENE_DEGRADATION removed due to gene number\n",
      "KEGG_SULFUR_METABOLISM removed due to gene number\n",
      "KEGG_RIBOSOME removed due to gene number\n",
      "KEGG_PROTEASOME removed due to gene number\n"
     ]
    }
   ],
   "source": [
    "# please keep the gene number setting same as GSEA command\n",
    "pruned_pathway_dict, gene_universe = prune_gmt(file_name='example/c2.cp.kegg.v2023.1.Hs.symbols.gmt', \n",
    "                                        out_file_name='example/c2.cp.kegg.v2023.1.Hs.symbols.cleaned.gmt', \n",
    "                                        expr_matrix_file='example/example_data.txt', \n",
    "                                        min_gene_num=15, max_gene_num=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0118934f",
   "metadata": {},
   "source": [
    "Fisher test takes a set gene of interest and a pathway file (dict only here). We'll perform this step for both top up and down-regulated genes, sorted by DESeq2 result p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "27924173",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# for extracting the top DEGs, we'll simply query the rank file here\n",
    "up_regulated_gene_set = extract_top_gene(rank_file='example/prerank.rnk', num_gene=200, direction=1)\n",
    "down_regulated_gene_set = extract_top_gene(rank_file='example/prerank.rnk', num_gene=200, direction=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5c3a30a1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start performing fisher test!\n",
      "KEGG_N_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_OTHER_GLYCAN_DEGRADATION\n",
      "KEGG_O_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\n",
      "KEGG_GLYCEROLIPID_METABOLISM\n",
      "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\n",
      "KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\n",
      "KEGG_ETHER_LIPID_METABOLISM\n",
      "KEGG_ARACHIDONIC_ACID_METABOLISM\n",
      "KEGG_LINOLEIC_ACID_METABOLISM\n",
      "KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\n",
      "KEGG_SPHINGOLIPID_METABOLISM\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\n",
      "KEGG_RIBOFLAVIN_METABOLISM\n",
      "KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\n",
      "KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\n",
      "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
      "KEGG_BASAL_TRANSCRIPTION_FACTORS\n",
      "KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\n",
      "KEGG_LYSOSOME\n",
      "KEGG_CARDIAC_MUSCLE_CONTRACTION\n",
      "KEGG_RENIN_ANGIOTENSIN_SYSTEM\n",
      "KEGG_TASTE_TRANSDUCTION\n",
      "KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\n",
      "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\n",
      "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\n",
      "KEGG_PRIMARY_IMMUNODEFICIENCY\n",
      "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\n",
      "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\n",
      "KEGG_DILATED_CARDIOMYOPATHY\n",
      "KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "KEGG_CITRATE_CYCLE_TCA_CYCLE\n",
      "KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
      "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
      "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\n",
      "KEGG_GALACTOSE_METABOLISM\n",
      "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\n",
      "KEGG_FATTY_ACID_METABOLISM\n",
      "KEGG_STEROID_BIOSYNTHESIS\n",
      "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
      "KEGG_STEROID_HORMONE_BIOSYNTHESIS\n",
      "KEGG_OXIDATIVE_PHOSPHORYLATION\n",
      "KEGG_PURINE_METABOLISM\n",
      "KEGG_PYRIMIDINE_METABOLISM\n",
      "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\n",
      "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\n",
      "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\n",
      "KEGG_LYSINE_DEGRADATION\n",
      "KEGG_ARGININE_AND_PROLINE_METABOLISM\n",
      "KEGG_HISTIDINE_METABOLISM\n",
      "KEGG_TYROSINE_METABOLISM\n",
      "KEGG_PHENYLALANINE_METABOLISM\n",
      "KEGG_TRYPTOPHAN_METABOLISM\n",
      "KEGG_BETA_ALANINE_METABOLISM\n",
      "KEGG_SELENOAMINO_ACID_METABOLISM\n",
      "KEGG_GLUTATHIONE_METABOLISM\n",
      "KEGG_STARCH_AND_SUCROSE_METABOLISM\n",
      "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\n",
      "KEGG_INOSITOL_PHOSPHATE_METABOLISM\n",
      "KEGG_PYRUVATE_METABOLISM\n",
      "KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\n",
      "KEGG_PROPANOATE_METABOLISM\n",
      "KEGG_BUTANOATE_METABOLISM\n",
      "KEGG_ONE_CARBON_POOL_BY_FOLATE\n",
      "KEGG_RETINOL_METABOLISM\n",
      "KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\n",
      "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\n",
      "KEGG_NITROGEN_METABOLISM\n",
      "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_OTHER_ENZYMES\n",
      "KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\n",
      "KEGG_ABC_TRANSPORTERS\n",
      "KEGG_RNA_DEGRADATION\n",
      "KEGG_RNA_POLYMERASE\n",
      "KEGG_DNA_REPLICATION\n",
      "KEGG_SPLICEOSOME\n",
      "KEGG_PROTEIN_EXPORT\n",
      "KEGG_PPAR_SIGNALING_PATHWAY\n",
      "KEGG_BASE_EXCISION_REPAIR\n",
      "KEGG_NUCLEOTIDE_EXCISION_REPAIR\n",
      "KEGG_MISMATCH_REPAIR\n",
      "KEGG_HOMOLOGOUS_RECOMBINATION\n",
      "KEGG_MAPK_SIGNALING_PATHWAY\n",
      "KEGG_ERBB_SIGNALING_PATHWAY\n",
      "KEGG_CALCIUM_SIGNALING_PATHWAY\n",
      "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\n",
      "KEGG_CHEMOKINE_SIGNALING_PATHWAY\n",
      "KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\n",
      "KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_CYCLE\n",
      "KEGG_OOCYTE_MEIOSIS\n",
      "KEGG_P53_SIGNALING_PATHWAY\n",
      "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\n",
      "KEGG_REGULATION_OF_AUTOPHAGY\n",
      "KEGG_ENDOCYTOSIS\n",
      "KEGG_PEROXISOME\n",
      "KEGG_MTOR_SIGNALING_PATHWAY\n",
      "KEGG_APOPTOSIS\n",
      "KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\n",
      "KEGG_WNT_SIGNALING_PATHWAY\n",
      "KEGG_DORSO_VENTRAL_AXIS_FORMATION\n",
      "KEGG_NOTCH_SIGNALING_PATHWAY\n",
      "KEGG_HEDGEHOG_SIGNALING_PATHWAY\n",
      "KEGG_TGF_BETA_SIGNALING_PATHWAY\n",
      "KEGG_AXON_GUIDANCE\n",
      "KEGG_VEGF_SIGNALING_PATHWAY\n",
      "KEGG_FOCAL_ADHESION\n",
      "KEGG_ECM_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_ADHESION_MOLECULES_CAMS\n",
      "KEGG_ADHERENS_JUNCTION\n",
      "KEGG_TIGHT_JUNCTION\n",
      "KEGG_GAP_JUNCTION\n",
      "KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\n",
      "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
      "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\n",
      "KEGG_JAK_STAT_SIGNALING_PATHWAY\n",
      "KEGG_HEMATOPOIETIC_CELL_LINEAGE\n",
      "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\n",
      "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\n",
      "KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\n",
      "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\n",
      "KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\n",
      "KEGG_LONG_TERM_POTENTIATION\n",
      "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\n",
      "KEGG_LONG_TERM_DEPRESSION\n",
      "KEGG_OLFACTORY_TRANSDUCTION\n",
      "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\n",
      "KEGG_INSULIN_SIGNALING_PATHWAY\n",
      "KEGG_GNRH_SIGNALING_PATHWAY\n",
      "KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\n",
      "KEGG_MELANOGENESIS\n",
      "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\n",
      "KEGG_TYPE_II_DIABETES_MELLITUS\n",
      "KEGG_TYPE_I_DIABETES_MELLITUS\n",
      "KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\n",
      "KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\n",
      "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\n",
      "KEGG_ALZHEIMERS_DISEASE\n",
      "KEGG_PARKINSONS_DISEASE\n",
      "KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\n",
      "KEGG_HUNTINGTONS_DISEASE\n",
      "KEGG_PRION_DISEASES\n",
      "KEGG_VIBRIO_CHOLERAE_INFECTION\n",
      "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\n",
      "KEGG_LEISHMANIA_INFECTION\n",
      "KEGG_PATHWAYS_IN_CANCER\n",
      "KEGG_COLORECTAL_CANCER\n",
      "KEGG_RENAL_CELL_CARCINOMA\n",
      "KEGG_PANCREATIC_CANCER\n",
      "KEGG_ENDOMETRIAL_CANCER\n",
      "KEGG_GLIOMA\n",
      "KEGG_PROSTATE_CANCER\n",
      "KEGG_THYROID_CANCER\n",
      "KEGG_BASAL_CELL_CARCINOMA\n",
      "KEGG_MELANOMA\n",
      "KEGG_BLADDER_CANCER\n",
      "KEGG_CHRONIC_MYELOID_LEUKEMIA\n",
      "KEGG_ACUTE_MYELOID_LEUKEMIA\n",
      "KEGG_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_NON_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_ASTHMA\n",
      "KEGG_AUTOIMMUNE_THYROID_DISEASE\n",
      "KEGG_ALLOGRAFT_REJECTION\n",
      "KEGG_GRAFT_VERSUS_HOST_DISEASE\n",
      "KEGG_VIRAL_MYOCARDITIS\n",
      "Results written to  example/fisher_up.txt\n",
      "********************\n",
      "Start performing fisher test!\n",
      "KEGG_N_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_OTHER_GLYCAN_DEGRADATION\n",
      "KEGG_O_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\n",
      "KEGG_GLYCEROLIPID_METABOLISM\n",
      "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\n",
      "KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\n",
      "KEGG_ETHER_LIPID_METABOLISM\n",
      "KEGG_ARACHIDONIC_ACID_METABOLISM\n",
      "KEGG_LINOLEIC_ACID_METABOLISM\n",
      "KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\n",
      "KEGG_SPHINGOLIPID_METABOLISM\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\n",
      "KEGG_RIBOFLAVIN_METABOLISM\n",
      "KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\n",
      "KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\n",
      "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
      "KEGG_BASAL_TRANSCRIPTION_FACTORS\n",
      "KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\n",
      "KEGG_LYSOSOME\n",
      "KEGG_CARDIAC_MUSCLE_CONTRACTION\n",
      "KEGG_RENIN_ANGIOTENSIN_SYSTEM\n",
      "KEGG_TASTE_TRANSDUCTION\n",
      "KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\n",
      "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\n",
      "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\n",
      "KEGG_PRIMARY_IMMUNODEFICIENCY\n",
      "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\n",
      "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\n",
      "KEGG_DILATED_CARDIOMYOPATHY\n",
      "KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "KEGG_CITRATE_CYCLE_TCA_CYCLE\n",
      "KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
      "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
      "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\n",
      "KEGG_GALACTOSE_METABOLISM\n",
      "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\n",
      "KEGG_FATTY_ACID_METABOLISM\n",
      "KEGG_STEROID_BIOSYNTHESIS\n",
      "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
      "KEGG_STEROID_HORMONE_BIOSYNTHESIS\n",
      "KEGG_OXIDATIVE_PHOSPHORYLATION\n",
      "KEGG_PURINE_METABOLISM\n",
      "KEGG_PYRIMIDINE_METABOLISM\n",
      "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\n",
      "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\n",
      "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\n",
      "KEGG_LYSINE_DEGRADATION\n",
      "KEGG_ARGININE_AND_PROLINE_METABOLISM\n",
      "KEGG_HISTIDINE_METABOLISM\n",
      "KEGG_TYROSINE_METABOLISM\n",
      "KEGG_PHENYLALANINE_METABOLISM\n",
      "KEGG_TRYPTOPHAN_METABOLISM\n",
      "KEGG_BETA_ALANINE_METABOLISM\n",
      "KEGG_SELENOAMINO_ACID_METABOLISM\n",
      "KEGG_GLUTATHIONE_METABOLISM\n",
      "KEGG_STARCH_AND_SUCROSE_METABOLISM\n",
      "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\n",
      "KEGG_INOSITOL_PHOSPHATE_METABOLISM\n",
      "KEGG_PYRUVATE_METABOLISM\n",
      "KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\n",
      "KEGG_PROPANOATE_METABOLISM\n",
      "KEGG_BUTANOATE_METABOLISM\n",
      "KEGG_ONE_CARBON_POOL_BY_FOLATE\n",
      "KEGG_RETINOL_METABOLISM\n",
      "KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\n",
      "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\n",
      "KEGG_NITROGEN_METABOLISM\n",
      "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_OTHER_ENZYMES\n",
      "KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\n",
      "KEGG_ABC_TRANSPORTERS\n",
      "KEGG_RNA_DEGRADATION\n",
      "KEGG_RNA_POLYMERASE\n",
      "KEGG_DNA_REPLICATION\n",
      "KEGG_SPLICEOSOME\n",
      "KEGG_PROTEIN_EXPORT\n",
      "KEGG_PPAR_SIGNALING_PATHWAY\n",
      "KEGG_BASE_EXCISION_REPAIR\n",
      "KEGG_NUCLEOTIDE_EXCISION_REPAIR\n",
      "KEGG_MISMATCH_REPAIR\n",
      "KEGG_HOMOLOGOUS_RECOMBINATION\n",
      "KEGG_MAPK_SIGNALING_PATHWAY\n",
      "KEGG_ERBB_SIGNALING_PATHWAY\n",
      "KEGG_CALCIUM_SIGNALING_PATHWAY\n",
      "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\n",
      "KEGG_CHEMOKINE_SIGNALING_PATHWAY\n",
      "KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\n",
      "KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_CYCLE\n",
      "KEGG_OOCYTE_MEIOSIS\n",
      "KEGG_P53_SIGNALING_PATHWAY\n",
      "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\n",
      "KEGG_REGULATION_OF_AUTOPHAGY\n",
      "KEGG_ENDOCYTOSIS\n",
      "KEGG_PEROXISOME\n",
      "KEGG_MTOR_SIGNALING_PATHWAY\n",
      "KEGG_APOPTOSIS\n",
      "KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\n",
      "KEGG_WNT_SIGNALING_PATHWAY\n",
      "KEGG_DORSO_VENTRAL_AXIS_FORMATION\n",
      "KEGG_NOTCH_SIGNALING_PATHWAY\n",
      "KEGG_HEDGEHOG_SIGNALING_PATHWAY\n",
      "KEGG_TGF_BETA_SIGNALING_PATHWAY\n",
      "KEGG_AXON_GUIDANCE\n",
      "KEGG_VEGF_SIGNALING_PATHWAY\n",
      "KEGG_FOCAL_ADHESION\n",
      "KEGG_ECM_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_ADHESION_MOLECULES_CAMS\n",
      "KEGG_ADHERENS_JUNCTION\n",
      "KEGG_TIGHT_JUNCTION\n",
      "KEGG_GAP_JUNCTION\n",
      "KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\n",
      "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
      "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\n",
      "KEGG_JAK_STAT_SIGNALING_PATHWAY\n",
      "KEGG_HEMATOPOIETIC_CELL_LINEAGE\n",
      "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\n",
      "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\n",
      "KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\n",
      "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\n",
      "KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\n",
      "KEGG_LONG_TERM_POTENTIATION\n",
      "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\n",
      "KEGG_LONG_TERM_DEPRESSION\n",
      "KEGG_OLFACTORY_TRANSDUCTION\n",
      "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\n",
      "KEGG_INSULIN_SIGNALING_PATHWAY\n",
      "KEGG_GNRH_SIGNALING_PATHWAY\n",
      "KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\n",
      "KEGG_MELANOGENESIS\n",
      "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\n",
      "KEGG_TYPE_II_DIABETES_MELLITUS\n",
      "KEGG_TYPE_I_DIABETES_MELLITUS\n",
      "KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\n",
      "KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\n",
      "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\n",
      "KEGG_ALZHEIMERS_DISEASE\n",
      "KEGG_PARKINSONS_DISEASE\n",
      "KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\n",
      "KEGG_HUNTINGTONS_DISEASE\n",
      "KEGG_PRION_DISEASES\n",
      "KEGG_VIBRIO_CHOLERAE_INFECTION\n",
      "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\n",
      "KEGG_LEISHMANIA_INFECTION\n",
      "KEGG_PATHWAYS_IN_CANCER\n",
      "KEGG_COLORECTAL_CANCER\n",
      "KEGG_RENAL_CELL_CARCINOMA\n",
      "KEGG_PANCREATIC_CANCER\n",
      "KEGG_ENDOMETRIAL_CANCER\n",
      "KEGG_GLIOMA\n",
      "KEGG_PROSTATE_CANCER\n",
      "KEGG_THYROID_CANCER\n",
      "KEGG_BASAL_CELL_CARCINOMA\n",
      "KEGG_MELANOMA\n",
      "KEGG_BLADDER_CANCER\n",
      "KEGG_CHRONIC_MYELOID_LEUKEMIA\n",
      "KEGG_ACUTE_MYELOID_LEUKEMIA\n",
      "KEGG_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_NON_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_ASTHMA\n",
      "KEGG_AUTOIMMUNE_THYROID_DISEASE\n",
      "KEGG_ALLOGRAFT_REJECTION\n",
      "KEGG_GRAFT_VERSUS_HOST_DISEASE\n",
      "KEGG_VIRAL_MYOCARDITIS\n",
      "Results written to  example/fisher_down.txt\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "# run fisher test for both sets of genes against the pathway file\n",
    "run_fisher_test(pathway_dict=pruned_pathway_dict, gene_set=up_regulated_gene_set, \n",
    "                gene_universe_num=len(gene_universe), out_file_name='example/fisher_up.txt')\n",
    "run_fisher_test(pathway_dict=pruned_pathway_dict, gene_set=down_regulated_gene_set, \n",
    "                gene_universe_num=len(gene_universe), out_file_name='example/fisher_down.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "751382f7",
   "metadata": {},
   "source": [
    "Running enrichr is similar to run fisher test, which requires a set of genes of interest and a pathway file.\\\n",
    "To caculate the final enrichment score, enrichr requires one step of permutation, **we recommend to set perm_num to 1000**.\\\n",
    "If the permutation file already exists, we'll use the existing one; if not, a new permutation will be performed. **NOTE**: the permutaiton step might take some time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "26b6c8cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "example/enrichr_kegg_permutation_1000.txt does not exist, initializing permutation\n",
      "Permutation 0\n",
      "Permutation 50\n",
      "Permutation 100\n",
      "Permutation 150\n",
      "Permutation 200\n",
      "Permutation 250\n",
      "Permutation 300\n",
      "Permutation 350\n",
      "Permutation 400\n",
      "Permutation 450\n",
      "Permutation 500\n",
      "Permutation 550\n",
      "Permutation 600\n",
      "Permutation 650\n",
      "Permutation 700\n",
      "Permutation 750\n",
      "Permutation 800\n",
      "Permutation 850\n",
      "Permutation 900\n",
      "Permutation 950\n",
      "Permutation result written to  example/enrichr_kegg_permutation_1000.txt\n",
      "Performing fisher test now...\n",
      "Results written to  example/enrichr_up.txt\n",
      "********************\n",
      "example/enrichr_kegg_permutation_1000.txt already exists, reading file now...\n",
      "Performing fisher test now...\n",
      "Results written to  example/enrichr_down.txt\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "run_enrichr(pathway_dict=pruned_pathway_dict, gene_set=up_regulated_gene_set,\n",
    "            gene_universe=gene_universe, out_file_name='example/enrichr_up.txt', \n",
    "            permutation_num=1000, permutation_file_name='example/enrichr_kegg_permutation_1000.txt')\n",
    "run_enrichr(pathway_dict=pruned_pathway_dict, gene_set=down_regulated_gene_set,\n",
    "            gene_universe=gene_universe, out_file_name='example/enrichr_down.txt', \n",
    "            permutation_num=1000, permutation_file_name='example/enrichr_kegg_permutation_1000.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dab8b90",
   "metadata": {},
   "source": [
    "After getting results from three underlying methods, the last step is to run PET. We'll perform this step for both up and down-regulated genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "74a85257",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating fisher rank...\n",
      "Calculating enrichr rank...\n",
      "Calculating GSEA rank...\n",
      "Ensembling results...\n",
      "Results written to  example/PET_output_enriched_poor_survival.txt\n",
      "Calculating fisher rank...\n",
      "Calculating enrichr rank...\n",
      "Calculating GSEA rank...\n",
      "Ensembling results...\n",
      "Results written to  example/PET_output_enriched_better_survival.txt\n"
     ]
    }
   ],
   "source": [
    "run_PET(fisher_result_file='example/fisher_up.txt', enrichr_result_file= 'example/enrichr_up.txt', \n",
    "        gsea_result_dir='example/example_test.GseaPreranked.1678566691197/', gsea_label='pos', \n",
    "        pathway_dict=pruned_pathway_dict, result_file='example/PET_output_enriched_poor_survival.txt')\n",
    "\n",
    "run_PET(fisher_result_file='example/fisher_down.txt', enrichr_result_file= 'example/enrichr_down.txt', \n",
    "        gsea_result_dir='example/example_test.GseaPreranked.1678566691197/', gsea_label='neg', \n",
    "        pathway_dict=pruned_pathway_dict, result_file='example/PET_output_enriched_better_survival.txt')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dde1977c",
   "metadata": {},
   "source": [
    "The results of PET is a tab-delimited text file, used could choose to sort the results based on their preferred criteria. Based on the benchmark we developed, the most reliable result would be sorting the pathways based on **average rank**, which by default generates the PET rank. At the same time, the significance PET FDR is also a strong indicator in the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e93de07",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
