{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "beb52269",
   "metadata": {},
   "source": [
    "This notebook is a tutorial for running PET.\n",
    "From the benchmark, we identified the optimal setting for running PET, which consists of 3 steps:\n",
    "1. perform differential expression analysis with DESeq2\n",
    "2. perform pathway analysis with underlying methods\n",
    "3. combine the result using PET\n",
    "\n",
    "File required:\n",
    "1. Data matrix, raw read count strongly recommended.\n",
    "2. Pathway file, in gmt format.\n",
    "\n",
    "Example data:\\\n",
    "We provided an example read count matrix text file and gmt pathway file (KEGG) as an example.\n",
    "\n",
    "Notes:\\\n",
    "Though template script and command for running GSEA and DESeq2 are provided, users should feel free to use any differential expression analysis methods and GSEA mode as wanted, as long as the output format is consistent for correctly parsing the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4c4e48a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from fisher_test import run_fisher_test\n",
    "from enrichr import run_enrichr\n",
    "from helper import *\n",
    "import os, sys\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "56a0face",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create result directory\n",
    "out_dir = 'example/'\n",
    "create_dir(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a55a109a",
   "metadata": {},
   "source": [
    "Step 1. \n",
    "Perform DEseq2 analysis, the template DESeq2 script is provided in deseq2_template.R.\\\n",
    "Note: **Raw read count** is required for DESeq2, first column as gene name, rest column as group1+group2 expression in order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ad8f7dbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DESeq2 script written to deseq_analysis.R\n"
     ]
    }
   ],
   "source": [
    "# helper function to format DEseq2 script, please specify the result file with .csv suffix\n",
    "format_deseq2_script(read_count_file_path='example/example_data.txt', sample_num=62, group1_id='poor_survival',\n",
    "                     group2_id='better_survival', group1_num=33, group2_num=29, \n",
    "                     result_file_path='example/deseq_result.csv', script_name='deseq_analysis.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ecbf785",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h         TCGA_AA_3850.Tumor.Rep244 TCGA_AA_3845.Tumor.Rep208\n",
      "TSPAN6                        2447                      2053\n",
      "TNMD                             9                         3\n",
      "DPM1                           490                       876\n",
      "SCYL3                          255                       299\n",
      "C1orf112                       164                       221\n",
      "FGR                            159                       152\n",
      "         TCGA_AA_3852.Tumor.Rep82 TCGA_G4_6302.Tumor.Rep159\n",
      "TSPAN6                       1454                      1822\n",
      "TNMD                            3                         2\n",
      "DPM1                          392                      1012\n",
      "SCYL3                         246                       468\n",
      "C1orf112                      147                       187\n",
      "FGR                           137                       534\n",
      "         TCGA_AA_3492.Tumor.Rep440 TCGA_G4_6626.Tumor.Rep431\n",
      "TSPAN6                        2356                      5642\n",
      "TNMD                             4                        20\n",
      "DPM1                           317                      4033\n",
      "SCYL3                          301                       766\n",
      "C1orf112                       396                       767\n",
      "FGR                            115                        68\n",
      "         TCGA_AA_3818.Tumor.Rep350 TCGA_AZ_6608.Tumor.Rep154\n",
      "TSPAN6                        6560                     11234\n",
      "TNMD                           103                        52\n",
      "DPM1                          1170                      4197\n",
      "SCYL3                          359                       705\n",
      "C1orf112                       306                       571\n",
      "FGR                             68                        42\n",
      "         TCGA_AA_A00N.Tumor.Rep442 TCGA_F4_6570.Tumor.Rep253\n",
      "TSPAN6                        1447                       548\n",
      "TNMD                             6                         0\n",
      "DPM1                           518                       817\n",
      "SCYL3                          215                       545\n",
      "C1orf112                       119                       319\n",
      "FGR                            159                      1007\n",
      "         TCGA_AZ_6599.Normal.Rep36 TCGA_AA_3489.Normal.Rep10\n",
      "TSPAN6                        7843                      6794\n",
      "TNMD                            15                        32\n",
      "DPM1                          1173                      1968\n",
      "SCYL3                          735                       959\n",
      "C1orf112                       294                       234\n",
      "FGR                            247                       662\n",
      "         TCGA_DM_A280.Tumor.Rep247 TCGA_CA_6718.Tumor.Rep5\n",
      "TSPAN6                        3648                    3100\n",
      "TNMD                            58                       1\n",
      "DPM1                          1259                    2108\n",
      "SCYL3                          430                     891\n",
      "C1orf112                       293                     795\n",
      "FGR                            134                     774\n",
      "         TCGA_DM_A1D7.Tumor.Rep280 TCGA_A6_4105.Tumor.Rep352\n",
      "TSPAN6                        9847                      9913\n",
      "TNMD                           209                        15\n",
      "DPM1                          4978                      4147\n",
      "SCYL3                          744                       679\n",
      "C1orf112                       571                       764\n",
      "FGR                             99                       477\n",
      "         TCGA_DM_A1D6.Tumor.Rep188 TCGA_AA_3715.Tumor.Rep319\n",
      "TSPAN6                        3919                       325\n",
      "TNMD                             0                         4\n",
      "DPM1                          4514                       172\n",
      "SCYL3                         1158                       175\n",
      "C1orf112                      1036                       221\n",
      "FGR                             70                       416\n",
      "         TCGA_AA_A02R.Tumor.Rep32 TCGA_DM_A0XD.Tumor.Rep252\n",
      "TSPAN6                       1040                      8183\n",
      "TNMD                            0                        59\n",
      "DPM1                          450                      3093\n",
      "SCYL3                         223                       467\n",
      "C1orf112                      174                       529\n",
      "FGR                           415                       112\n",
      "         TCGA_AZ_6603.Tumor.Rep477 TCGA_A6_2686.Tumor.Rep445\n",
      "TSPAN6                       21043                      7111\n",
      "TNMD                           155                        39\n",
      "DPM1                          5589                      2102\n",
      "SCYL3                          793                       860\n",
      "C1orf112                       532                       423\n",
      "FGR                            434                       885\n",
      "         TCGA_AA_A024.Tumor.Rep392 TCGA_A6_2676.Tumor.Rep121\n",
      "TSPAN6                        2746                      1601\n",
      "TNMD                            70                         0\n",
      "DPM1                           356                       610\n",
      "SCYL3                          214                       185\n",
      "C1orf112                        72                       233\n",
      "FGR                             23                       381\n",
      "         TCGA_DM_A1DB.Tumor.Rep376 TCGA_CK_5912.Tumor.Rep329\n",
      "TSPAN6                       33725                     12484\n",
      "TNMD                           224                        40\n",
      "DPM1                         10982                      4011\n",
      "SCYL3                          946                       856\n",
      "C1orf112                      1070                       737\n",
      "FGR                            213                       128\n",
      "         TCGA_AZ_6598.Tumor.Rep404 TCGA_DM_A28G.Tumor.Rep157\n",
      "TSPAN6                        6198                      8013\n",
      "TNMD                             0                        99\n",
      "DPM1                          1666                      1894\n",
      "SCYL3                          434                      1118\n",
      "C1orf112                       303                       410\n",
      "FGR                            361                        90\n",
      "         TCGA_AZ_5403.Tumor.Rep453 TCGA_CK_4951.Tumor.Rep295\n",
      "TSPAN6                       16799                      6513\n",
      "TNMD                            50                         8\n",
      "DPM1                          5111                       959\n",
      "SCYL3                          654                       307\n",
      "C1orf112                       567                       246\n",
      "FGR                            281                       324\n",
      "         TCGA_DM_A28C.Tumor.Rep136 TCGA_DM_A1D4.Tumor.Rep347\n",
      "TSPAN6                        4912                      6611\n",
      "TNMD                             3                        36\n",
      "DPM1                          1344                      3725\n",
      "SCYL3                         1245                      1188\n",
      "C1orf112                       690                       741\n",
      "FGR                             23                        89\n",
      "         TCGA_AZ_6601.Tumor.Rep438 TCGA_AA_3693.Tumor.Rep205\n",
      "TSPAN6                        4335                      8210\n",
      "TNMD                             7                        21\n",
      "DPM1                          1640                      3753\n",
      "SCYL3                          675                       615\n",
      "C1orf112                       685                       350\n",
      "FGR                           1456                        67\n",
      "         TCGA_AA_3672.Tumor.Rep18 TCGA_AA_3681.Tumor.Rep436\n",
      "TSPAN6                        302                      1467\n",
      "TNMD                            3                         6\n",
      "DPM1                           83                       821\n",
      "SCYL3                          71                       245\n",
      "C1orf112                       62                       172\n",
      "FGR                           172                       117\n",
      "         TCGA_AA_3662.Tumor.Rep172 TCGA_D5_6922.Tumor.Rep152\n",
      "TSPAN6                       10570                     11697\n",
      "TNMD                            57                        37\n",
      "DPM1                          1973                      2843\n",
      "SCYL3                          799                      1259\n",
      "C1orf112                       491                       869\n",
      "FGR                            849                       438\n",
      "         TCGA_D5_6931.Tumor.Rep330 TCGA_CM_6680.Tumor.Rep451\n",
      "TSPAN6                        5040                      7398\n",
      "TNMD                            40                         5\n",
      "DPM1                          1461                      1544\n",
      "SCYL3                          796                       889\n",
      "C1orf112                       757                       436\n",
      "FGR                            305                       352\n",
      "         TCGA_AA_3844.Tumor.Rep427 TCGA_AA_A01C.Tumor.Rep174\n",
      "TSPAN6                        4929                      1990\n",
      "TNMD                             8                         8\n",
      "DPM1                          1089                       598\n",
      "SCYL3                          319                        81\n",
      "C1orf112                       252                        75\n",
      "FGR                             83                        34\n",
      "         TCGA_D5_6535.Tumor.Rep70 TCGA_D5_6538.Tumor.Rep284\n",
      "TSPAN6                       4721                     11094\n",
      "TNMD                           58                        11\n",
      "DPM1                         2203                      2392\n",
      "SCYL3                        1001                      1099\n",
      "C1orf112                      689                       404\n",
      "FGR                           278                        84\n",
      "         TCGA_AA_A00J.Tumor.Rep403 TCGA_AA_3688.Tumor.Rep231\n",
      "TSPAN6                        1210                     11751\n",
      "TNMD                             3                        26\n",
      "DPM1                           641                      2494\n",
      "SCYL3                          311                       531\n",
      "C1orf112                       182                       632\n",
      "FGR                            130                        85\n",
      "         TCGA_AA_3562.Tumor.Rep405 TCGA_CK_5914.Tumor.Rep91\n",
      "TSPAN6                       13065                    10439\n",
      "TNMD                            37                       61\n",
      "DPM1                          1760                     8789\n",
      "SCYL3                          297                      700\n",
      "C1orf112                       289                      794\n",
      "FGR                            164                      167\n",
      "         TCGA_A6_5664.Tumor.Rep471 TCGA_AA_3941.Tumor.Rep291\n",
      "TSPAN6                        3670                      2984\n",
      "TNMD                             9                         2\n",
      "DPM1                          1244                      1128\n",
      "SCYL3                          905                       328\n",
      "C1orf112                       477                       142\n",
      "FGR                            678                        51\n",
      "         TCGA_AD_6895.Tumor.Rep79 TCGA_AA_A01X.Tumor.Rep328\n",
      "TSPAN6                        797                      2823\n",
      "TNMD                            0                         2\n",
      "DPM1                         1513                       568\n",
      "SCYL3                         656                       790\n",
      "C1orf112                      333                       397\n",
      "FGR                           316                        58\n",
      "         TCGA_AA_3949.Tumor.Rep19 TCGA_AA_3982.Tumor.Rep406\n",
      "TSPAN6                       1786                      5979\n",
      "TNMD                            5                        36\n",
      "DPM1                          698                      1998\n",
      "SCYL3                         214                       323\n",
      "C1orf112                      103                       207\n",
      "FGR                           291                       156\n",
      "         TCGA_AA_A00O.Tumor.Rep183 TCGA_CM_5341.Tumor.Rep120\n",
      "TSPAN6                        1595                      9167\n",
      "TNMD                             0                        23\n",
      "DPM1                          1609                      3228\n",
      "SCYL3                          365                       550\n",
      "C1orf112                       104                       287\n",
      "FGR                            185                       300\n",
      "         TCGA_AZ_4615.Tumor.Rep197 TCGA_G4_6314.Tumor.Rep294\n",
      "TSPAN6                        2349                      8201\n",
      "TNMD                             1                        36\n",
      "DPM1                           960                      2264\n",
      "SCYL3                          303                       707\n",
      "C1orf112                       173                       575\n",
      "FGR                            190                       323\n",
      "         TCGA_G4_6311.Tumor.Rep387 TCGA_A6_2672.Tumor.Rep367\n",
      "TSPAN6                        5085                       747\n",
      "TNMD                             2                         3\n",
      "DPM1                          2981                       265\n",
      "SCYL3                          658                       527\n",
      "C1orf112                       845                       249\n",
      "FGR                            459                       340\n",
      "         TCGA_DM_A1HA.Tumor.Rep410 TCGA_DM_A1HB.Tumor.Rep417\n",
      "TSPAN6                        2519                      2463\n",
      "TNMD                             1                         0\n",
      "DPM1                          3555                      2408\n",
      "SCYL3                          818                       776\n",
      "C1orf112                       386                       430\n",
      "FGR                             71                       164\n",
      "\u001b[?25h [1] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      " [5] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      " [9] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[13] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[17] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[21] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[25] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[29] poor_survival   poor_survival   poor_survival   poor_survival  \n",
      "[33] poor_survival   better_survival better_survival better_survival\n",
      "[37] better_survival better_survival better_survival better_survival\n",
      "[41] better_survival better_survival better_survival better_survival\n",
      "[45] better_survival better_survival better_survival better_survival\n",
      "[49] better_survival better_survival better_survival better_survival\n",
      "[53] better_survival better_survival better_survival better_survival\n",
      "[57] better_survival better_survival better_survival better_survival\n",
      "[61] better_survival better_survival\n",
      "Levels: better_survival poor_survival\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hclass: DESeqDataSet \n",
      "dim: 53657 62 \n",
      "metadata(1): version\n",
      "assays(1): counts\n",
      "rownames(53657): TSPAN6 TNMD ... LINC01144 AC007389.5\n",
      "rowData names(0):\n",
      "colnames(62): TCGA_AA_3850.Tumor.Rep244 TCGA_AA_3845.Tumor.Rep208 ...\n",
      "  TCGA_DM_A1HA.Tumor.Rep410 TCGA_DM_A1HB.Tumor.Rep417\n",
      "colData names(1): condition\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "-- replacing outliers and refitting for 5411 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "estimating dispersions\n",
      "fitting model and testing\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "# run DESeq2 script\n",
    "! Rscript deseq_analysis.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ca355d66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract DESeq2 result, CSV file only, to .rnk file for GSEA, please do check the direction of DESeq2 result (log2FoldChange column)\n",
    "# if log2FoldChange>=0 means up-regulation, keep direction=1; otherwise, pass direction=-1\n",
    "generate_rank_file(deseq_result_file='example/deseq_result.csv', out_file='example/prerank.rnk', direction=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5156a2a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using system JDK.\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "494      [INFO  ] - Parameters passing to GSEAPreranked.main:\n",
      "496      [INFO  ] - rnk\texample/prerank.rnk\n",
      "496      [INFO  ] - gmx\texample/c2.cp.kegg.v2023.1.Hs.symbols.gmt\n",
      "496      [INFO  ] - rpt_label\texample_test\n",
      "496      [INFO  ] - collapse\tNo_Collapse\n",
      "496      [INFO  ] - zip_report\tfalse\n",
      "496      [INFO  ] - gui\tfalse\n",
      "496      [INFO  ] - out\texample/gsea_out_example/\n",
      "496      [INFO  ] - mode\tMax_probe\n",
      "496      [INFO  ] - norm\tmeandiv\n",
      "496      [INFO  ] - nperm\t1000\n",
      "496      [INFO  ] - scoring_scheme\tweighted\n",
      "496      [INFO  ] - include_only_symbols\ttrue\n",
      "496      [INFO  ] - make_sets\ttrue\n",
      "496      [INFO  ] - plot_top_x\t5\n",
      "496      [INFO  ] - rnd_seed\ttimestamp\n",
      "496      [INFO  ] - create_svgs\tfalse\n",
      "496      [INFO  ] - set_max\t500\n",
      "497      [INFO  ] - set_min\t15\n",
      "638      [INFO  ] - Made Vdb dir JIT: /Users/luopin/GSEA_test/ENCODE_RNA-seq/PET/mar10\n",
      "656      [INFO  ] - Begun importing: RankedList from: example/prerank.rnk\n",
      "1130     [INFO  ] - Your current version of GSEA is 4.1.0. A newer version (4.3.2) is available.\n",
      "To update, please download from http://www.gsea-msigdb.org/gsea/gsea/downloads.jsp\n",
      "(build date: Mon, 3 Oct 2022 16:56:43 -0400)\n",
      " Better support for our new Mouse MSigDB release\n",
      "1130     [INFO  ] - Note: GenePattern users should update through GenePattern.\n",
      "1171     [INFO  ] - No ranked list collapsing was done .. using original as is\n",
      "to parse>example/c2.cp.kegg.v2023.1.Hs.symbols.gmt< got: [example/c2.cp.kegg.v2023.1.Hs.symbols.gmt]\n",
      "1172     [INFO  ] - Begun importing: GeneSetMatrix from: c2.cp.kegg.v2023.1.Hs.symbols.gmt\n",
      "1198     [INFO  ] - Got gsets: 186 now preprocessing them ... min: 15 max: 500\n",
      "1213     [INFO  ] - Done preproc for smaller than: 15\n",
      "1326     [INFO  ] - Timestamp used as the random seed: 1678483249994\n",
      "shuffleGeneSet for GeneSet 1/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 6/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 11/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 16/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 21/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 26/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 31/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 36/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 41/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 46/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 51/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 56/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 61/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 66/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 71/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 76/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 81/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 86/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 91/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 96/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 101/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 106/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 111/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 116/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 121/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 126/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 131/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 136/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 141/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 146/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 151/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 156/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 161/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 166/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 171/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "shuffleGeneSet for GeneSet 176/176 nperm: 1000\n",
      "GeneSetCohorted: 501 / 1000\n",
      "GeneSetCohorted_scored: 501 / 1000\n",
      "29203    [INFO  ] - Creating marker selection reports ...\n",
      "29287    [INFO  ] - Creating FDR reports ...\n",
      "31013    [INFO  ] - Done FDR reports for positive phenotype\n",
      "31564    [INFO  ] - Done FDR reports for negative phenotype\n",
      "31659    [INFO  ] - Creating global reports ...\n",
      "31659    [INFO  ] - Done all reports!!\n",
      "Time taken: 30 secs\n"
     ]
    }
   ],
   "source": [
    "# To run GSEA, please download GSEA command line tool from here: http://www.gsea-msigdb.org/gsea/downloads.jsp\n",
    "# please change the gsea-cli.sh pathway as needed\n",
    "! ~/GSEA_test/GSEA_cmd/gsea-cli.sh GSEAPreranked -gmx example/c2.cp.kegg.v2023.1.Hs.symbols.gmt -collapse No_Collapse -mode Max_probe -norm meandiv -nperm 1000 -rnk example/prerank.rnk  -scoring_scheme weighted -rpt_label example_test   -create_svgs false -include_only_symbols true -make_sets true -plot_top_x 5 -rnd_seed timestamp -set_max 500 -set_min 15 -zip_report false -out example/gsea_out_example/  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40b39f1a",
   "metadata": {},
   "source": [
    "Before running other methods, we shall prune the pathways, which will drop pathways with gene_num > max_num and gene_num < min_num and also remove any gene that's not present in the expression matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1c0c4fa9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53657 genes in the expression matrix\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES removed due to gene number\n",
      "KEGG_NON_HOMOLOGOUS_END_JOINING removed due to gene number\n",
      "KEGG_CIRCADIAN_RHYTHM_MAMMAL removed due to gene number\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS removed due to gene number\n",
      "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM removed due to gene number\n",
      "KEGG_FOLATE_BIOSYNTHESIS removed due to gene number\n",
      "KEGG_LIMONENE_AND_PINENE_DEGRADATION removed due to gene number\n",
      "KEGG_SULFUR_METABOLISM removed due to gene number\n",
      "KEGG_RIBOSOME removed due to gene number\n",
      "KEGG_PROTEASOME removed due to gene number\n"
     ]
    }
   ],
   "source": [
    "# please keep the gene number setting same as GSEA command\n",
    "pruned_pathway_dict, gene_universe = prune_gmt(file_name='example/c2.cp.kegg.v2023.1.Hs.symbols.gmt', \n",
    "                                        out_file_name='example/c2.cp.kegg.v2023.1.Hs.symbols.cleaned.gmt', \n",
    "                                        expr_matrix_file='example/example_data.txt', \n",
    "                                        min_gene_num=15, max_gene_num=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a046d05f",
   "metadata": {},
   "source": [
    "Fisher test takes a set gene of interest and a pathway file (dict only here). We'll perform this step for both top up and down-regulated genes, sorted by DESeq2 result p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "455eb2c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for extracting the top DEGs, we'll simply query the rank file here\n",
    "up_regulated_gene_set = extract_top_gene(rank_file='example/prerank.rnk', num_gene=200, direction=1)\n",
    "down_regulated_gene_set = extract_top_gene(rank_file='example/prerank.rnk', num_gene=200, direction=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8be90959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start performing fisher test!\n",
      "KEGG_N_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_OTHER_GLYCAN_DEGRADATION\n",
      "KEGG_O_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\n",
      "KEGG_GLYCEROLIPID_METABOLISM\n",
      "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\n",
      "KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\n",
      "KEGG_ETHER_LIPID_METABOLISM\n",
      "KEGG_ARACHIDONIC_ACID_METABOLISM\n",
      "KEGG_LINOLEIC_ACID_METABOLISM\n",
      "KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\n",
      "KEGG_SPHINGOLIPID_METABOLISM\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\n",
      "KEGG_RIBOFLAVIN_METABOLISM\n",
      "KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\n",
      "KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\n",
      "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
      "KEGG_BASAL_TRANSCRIPTION_FACTORS\n",
      "KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\n",
      "KEGG_LYSOSOME\n",
      "KEGG_CARDIAC_MUSCLE_CONTRACTION\n",
      "KEGG_RENIN_ANGIOTENSIN_SYSTEM\n",
      "KEGG_TASTE_TRANSDUCTION\n",
      "KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\n",
      "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\n",
      "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\n",
      "KEGG_PRIMARY_IMMUNODEFICIENCY\n",
      "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\n",
      "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\n",
      "KEGG_DILATED_CARDIOMYOPATHY\n",
      "KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "KEGG_CITRATE_CYCLE_TCA_CYCLE\n",
      "KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
      "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
      "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\n",
      "KEGG_GALACTOSE_METABOLISM\n",
      "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\n",
      "KEGG_FATTY_ACID_METABOLISM\n",
      "KEGG_STEROID_BIOSYNTHESIS\n",
      "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
      "KEGG_STEROID_HORMONE_BIOSYNTHESIS\n",
      "KEGG_OXIDATIVE_PHOSPHORYLATION\n",
      "KEGG_PURINE_METABOLISM\n",
      "KEGG_PYRIMIDINE_METABOLISM\n",
      "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\n",
      "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\n",
      "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\n",
      "KEGG_LYSINE_DEGRADATION\n",
      "KEGG_ARGININE_AND_PROLINE_METABOLISM\n",
      "KEGG_HISTIDINE_METABOLISM\n",
      "KEGG_TYROSINE_METABOLISM\n",
      "KEGG_PHENYLALANINE_METABOLISM\n",
      "KEGG_TRYPTOPHAN_METABOLISM\n",
      "KEGG_BETA_ALANINE_METABOLISM\n",
      "KEGG_SELENOAMINO_ACID_METABOLISM\n",
      "KEGG_GLUTATHIONE_METABOLISM\n",
      "KEGG_STARCH_AND_SUCROSE_METABOLISM\n",
      "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\n",
      "KEGG_INOSITOL_PHOSPHATE_METABOLISM\n",
      "KEGG_PYRUVATE_METABOLISM\n",
      "KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\n",
      "KEGG_PROPANOATE_METABOLISM\n",
      "KEGG_BUTANOATE_METABOLISM\n",
      "KEGG_ONE_CARBON_POOL_BY_FOLATE\n",
      "KEGG_RETINOL_METABOLISM\n",
      "KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\n",
      "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\n",
      "KEGG_NITROGEN_METABOLISM\n",
      "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_OTHER_ENZYMES\n",
      "KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\n",
      "KEGG_ABC_TRANSPORTERS\n",
      "KEGG_RNA_DEGRADATION\n",
      "KEGG_RNA_POLYMERASE\n",
      "KEGG_DNA_REPLICATION\n",
      "KEGG_SPLICEOSOME\n",
      "KEGG_PROTEIN_EXPORT\n",
      "KEGG_PPAR_SIGNALING_PATHWAY\n",
      "KEGG_BASE_EXCISION_REPAIR\n",
      "KEGG_NUCLEOTIDE_EXCISION_REPAIR\n",
      "KEGG_MISMATCH_REPAIR\n",
      "KEGG_HOMOLOGOUS_RECOMBINATION\n",
      "KEGG_MAPK_SIGNALING_PATHWAY\n",
      "KEGG_ERBB_SIGNALING_PATHWAY\n",
      "KEGG_CALCIUM_SIGNALING_PATHWAY\n",
      "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\n",
      "KEGG_CHEMOKINE_SIGNALING_PATHWAY\n",
      "KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\n",
      "KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_CYCLE\n",
      "KEGG_OOCYTE_MEIOSIS\n",
      "KEGG_P53_SIGNALING_PATHWAY\n",
      "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\n",
      "KEGG_REGULATION_OF_AUTOPHAGY\n",
      "KEGG_ENDOCYTOSIS\n",
      "KEGG_PEROXISOME\n",
      "KEGG_MTOR_SIGNALING_PATHWAY\n",
      "KEGG_APOPTOSIS\n",
      "KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\n",
      "KEGG_WNT_SIGNALING_PATHWAY\n",
      "KEGG_DORSO_VENTRAL_AXIS_FORMATION\n",
      "KEGG_NOTCH_SIGNALING_PATHWAY\n",
      "KEGG_HEDGEHOG_SIGNALING_PATHWAY\n",
      "KEGG_TGF_BETA_SIGNALING_PATHWAY\n",
      "KEGG_AXON_GUIDANCE\n",
      "KEGG_VEGF_SIGNALING_PATHWAY\n",
      "KEGG_FOCAL_ADHESION\n",
      "KEGG_ECM_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_ADHESION_MOLECULES_CAMS\n",
      "KEGG_ADHERENS_JUNCTION\n",
      "KEGG_TIGHT_JUNCTION\n",
      "KEGG_GAP_JUNCTION\n",
      "KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\n",
      "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
      "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\n",
      "KEGG_JAK_STAT_SIGNALING_PATHWAY\n",
      "KEGG_HEMATOPOIETIC_CELL_LINEAGE\n",
      "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\n",
      "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\n",
      "KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\n",
      "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\n",
      "KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\n",
      "KEGG_LONG_TERM_POTENTIATION\n",
      "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\n",
      "KEGG_LONG_TERM_DEPRESSION\n",
      "KEGG_OLFACTORY_TRANSDUCTION\n",
      "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\n",
      "KEGG_INSULIN_SIGNALING_PATHWAY\n",
      "KEGG_GNRH_SIGNALING_PATHWAY\n",
      "KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\n",
      "KEGG_MELANOGENESIS\n",
      "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\n",
      "KEGG_TYPE_II_DIABETES_MELLITUS\n",
      "KEGG_TYPE_I_DIABETES_MELLITUS\n",
      "KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\n",
      "KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\n",
      "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\n",
      "KEGG_ALZHEIMERS_DISEASE\n",
      "KEGG_PARKINSONS_DISEASE\n",
      "KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\n",
      "KEGG_HUNTINGTONS_DISEASE\n",
      "KEGG_PRION_DISEASES\n",
      "KEGG_VIBRIO_CHOLERAE_INFECTION\n",
      "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\n",
      "KEGG_LEISHMANIA_INFECTION\n",
      "KEGG_PATHWAYS_IN_CANCER\n",
      "KEGG_COLORECTAL_CANCER\n",
      "KEGG_RENAL_CELL_CARCINOMA\n",
      "KEGG_PANCREATIC_CANCER\n",
      "KEGG_ENDOMETRIAL_CANCER\n",
      "KEGG_GLIOMA\n",
      "KEGG_PROSTATE_CANCER\n",
      "KEGG_THYROID_CANCER\n",
      "KEGG_BASAL_CELL_CARCINOMA\n",
      "KEGG_MELANOMA\n",
      "KEGG_BLADDER_CANCER\n",
      "KEGG_CHRONIC_MYELOID_LEUKEMIA\n",
      "KEGG_ACUTE_MYELOID_LEUKEMIA\n",
      "KEGG_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_NON_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_ASTHMA\n",
      "KEGG_AUTOIMMUNE_THYROID_DISEASE\n",
      "KEGG_ALLOGRAFT_REJECTION\n",
      "KEGG_GRAFT_VERSUS_HOST_DISEASE\n",
      "KEGG_VIRAL_MYOCARDITIS\n",
      "Results written to  example/fisher_up.txt\n",
      "********************\n",
      "Start performing fisher test!\n",
      "KEGG_N_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_OTHER_GLYCAN_DEGRADATION\n",
      "KEGG_O_GLYCAN_BIOSYNTHESIS\n",
      "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\n",
      "KEGG_GLYCEROLIPID_METABOLISM\n",
      "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\n",
      "KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\n",
      "KEGG_ETHER_LIPID_METABOLISM\n",
      "KEGG_ARACHIDONIC_ACID_METABOLISM\n",
      "KEGG_LINOLEIC_ACID_METABOLISM\n",
      "KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\n",
      "KEGG_SPHINGOLIPID_METABOLISM\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\n",
      "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\n",
      "KEGG_RIBOFLAVIN_METABOLISM\n",
      "KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\n",
      "KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\n",
      "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
      "KEGG_BASAL_TRANSCRIPTION_FACTORS\n",
      "KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\n",
      "KEGG_LYSOSOME\n",
      "KEGG_CARDIAC_MUSCLE_CONTRACTION\n",
      "KEGG_RENIN_ANGIOTENSIN_SYSTEM\n",
      "KEGG_TASTE_TRANSDUCTION\n",
      "KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\n",
      "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\n",
      "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\n",
      "KEGG_PRIMARY_IMMUNODEFICIENCY\n",
      "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\n",
      "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\n",
      "KEGG_DILATED_CARDIOMYOPATHY\n",
      "KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
      "KEGG_CITRATE_CYCLE_TCA_CYCLE\n",
      "KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
      "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
      "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\n",
      "KEGG_GALACTOSE_METABOLISM\n",
      "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\n",
      "KEGG_FATTY_ACID_METABOLISM\n",
      "KEGG_STEROID_BIOSYNTHESIS\n",
      "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
      "KEGG_STEROID_HORMONE_BIOSYNTHESIS\n",
      "KEGG_OXIDATIVE_PHOSPHORYLATION\n",
      "KEGG_PURINE_METABOLISM\n",
      "KEGG_PYRIMIDINE_METABOLISM\n",
      "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\n",
      "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\n",
      "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\n",
      "KEGG_LYSINE_DEGRADATION\n",
      "KEGG_ARGININE_AND_PROLINE_METABOLISM\n",
      "KEGG_HISTIDINE_METABOLISM\n",
      "KEGG_TYROSINE_METABOLISM\n",
      "KEGG_PHENYLALANINE_METABOLISM\n",
      "KEGG_TRYPTOPHAN_METABOLISM\n",
      "KEGG_BETA_ALANINE_METABOLISM\n",
      "KEGG_SELENOAMINO_ACID_METABOLISM\n",
      "KEGG_GLUTATHIONE_METABOLISM\n",
      "KEGG_STARCH_AND_SUCROSE_METABOLISM\n",
      "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\n",
      "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\n",
      "KEGG_INOSITOL_PHOSPHATE_METABOLISM\n",
      "KEGG_PYRUVATE_METABOLISM\n",
      "KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\n",
      "KEGG_PROPANOATE_METABOLISM\n",
      "KEGG_BUTANOATE_METABOLISM\n",
      "KEGG_ONE_CARBON_POOL_BY_FOLATE\n",
      "KEGG_RETINOL_METABOLISM\n",
      "KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\n",
      "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\n",
      "KEGG_NITROGEN_METABOLISM\n",
      "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_CYTOCHROME_P450\n",
      "KEGG_DRUG_METABOLISM_OTHER_ENZYMES\n",
      "KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\n",
      "KEGG_ABC_TRANSPORTERS\n",
      "KEGG_RNA_DEGRADATION\n",
      "KEGG_RNA_POLYMERASE\n",
      "KEGG_DNA_REPLICATION\n",
      "KEGG_SPLICEOSOME\n",
      "KEGG_PROTEIN_EXPORT\n",
      "KEGG_PPAR_SIGNALING_PATHWAY\n",
      "KEGG_BASE_EXCISION_REPAIR\n",
      "KEGG_NUCLEOTIDE_EXCISION_REPAIR\n",
      "KEGG_MISMATCH_REPAIR\n",
      "KEGG_HOMOLOGOUS_RECOMBINATION\n",
      "KEGG_MAPK_SIGNALING_PATHWAY\n",
      "KEGG_ERBB_SIGNALING_PATHWAY\n",
      "KEGG_CALCIUM_SIGNALING_PATHWAY\n",
      "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\n",
      "KEGG_CHEMOKINE_SIGNALING_PATHWAY\n",
      "KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\n",
      "KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_CYCLE\n",
      "KEGG_OOCYTE_MEIOSIS\n",
      "KEGG_P53_SIGNALING_PATHWAY\n",
      "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\n",
      "KEGG_REGULATION_OF_AUTOPHAGY\n",
      "KEGG_ENDOCYTOSIS\n",
      "KEGG_PEROXISOME\n",
      "KEGG_MTOR_SIGNALING_PATHWAY\n",
      "KEGG_APOPTOSIS\n",
      "KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\n",
      "KEGG_WNT_SIGNALING_PATHWAY\n",
      "KEGG_DORSO_VENTRAL_AXIS_FORMATION\n",
      "KEGG_NOTCH_SIGNALING_PATHWAY\n",
      "KEGG_HEDGEHOG_SIGNALING_PATHWAY\n",
      "KEGG_TGF_BETA_SIGNALING_PATHWAY\n",
      "KEGG_AXON_GUIDANCE\n",
      "KEGG_VEGF_SIGNALING_PATHWAY\n",
      "KEGG_FOCAL_ADHESION\n",
      "KEGG_ECM_RECEPTOR_INTERACTION\n",
      "KEGG_CELL_ADHESION_MOLECULES_CAMS\n",
      "KEGG_ADHERENS_JUNCTION\n",
      "KEGG_TIGHT_JUNCTION\n",
      "KEGG_GAP_JUNCTION\n",
      "KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\n",
      "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
      "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\n",
      "KEGG_JAK_STAT_SIGNALING_PATHWAY\n",
      "KEGG_HEMATOPOIETIC_CELL_LINEAGE\n",
      "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\n",
      "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
      "KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\n",
      "KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\n",
      "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\n",
      "KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\n",
      "KEGG_LONG_TERM_POTENTIATION\n",
      "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\n",
      "KEGG_LONG_TERM_DEPRESSION\n",
      "KEGG_OLFACTORY_TRANSDUCTION\n",
      "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\n",
      "KEGG_INSULIN_SIGNALING_PATHWAY\n",
      "KEGG_GNRH_SIGNALING_PATHWAY\n",
      "KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\n",
      "KEGG_MELANOGENESIS\n",
      "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\n",
      "KEGG_TYPE_II_DIABETES_MELLITUS\n",
      "KEGG_TYPE_I_DIABETES_MELLITUS\n",
      "KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\n",
      "KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\n",
      "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\n",
      "KEGG_ALZHEIMERS_DISEASE\n",
      "KEGG_PARKINSONS_DISEASE\n",
      "KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\n",
      "KEGG_HUNTINGTONS_DISEASE\n",
      "KEGG_PRION_DISEASES\n",
      "KEGG_VIBRIO_CHOLERAE_INFECTION\n",
      "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\n",
      "KEGG_LEISHMANIA_INFECTION\n",
      "KEGG_PATHWAYS_IN_CANCER\n",
      "KEGG_COLORECTAL_CANCER\n",
      "KEGG_RENAL_CELL_CARCINOMA\n",
      "KEGG_PANCREATIC_CANCER\n",
      "KEGG_ENDOMETRIAL_CANCER\n",
      "KEGG_GLIOMA\n",
      "KEGG_PROSTATE_CANCER\n",
      "KEGG_THYROID_CANCER\n",
      "KEGG_BASAL_CELL_CARCINOMA\n",
      "KEGG_MELANOMA\n",
      "KEGG_BLADDER_CANCER\n",
      "KEGG_CHRONIC_MYELOID_LEUKEMIA\n",
      "KEGG_ACUTE_MYELOID_LEUKEMIA\n",
      "KEGG_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_NON_SMALL_CELL_LUNG_CANCER\n",
      "KEGG_ASTHMA\n",
      "KEGG_AUTOIMMUNE_THYROID_DISEASE\n",
      "KEGG_ALLOGRAFT_REJECTION\n",
      "KEGG_GRAFT_VERSUS_HOST_DISEASE\n",
      "KEGG_VIRAL_MYOCARDITIS\n",
      "Results written to  example/fisher_down.txt\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "# run fisher test for both sets of genes against the pathway file\n",
    "run_fisher_test(pathway_dict=pruned_pathway_dict, gene_set=up_regulated_gene_set, \n",
    "                gene_universe_num=len(gene_universe), out_file_name='example/fisher_up.txt')\n",
    "run_fisher_test(pathway_dict=pruned_pathway_dict, gene_set=down_regulated_gene_set, \n",
    "                gene_universe_num=len(gene_universe), out_file_name='example/fisher_down.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd16b2a",
   "metadata": {},
   "source": [
    "Running enrichr is similar to run fisher test, which requires a set of genes of interest and a pathway file.\\\n",
    "To caculate the final enrichment score, enrichr requires one step of permutation, **we recommend to set perm_num to 1000**.\\\n",
    "If the permutation file already exists, we'll use the existing one; if not, a new permutation will be performed. **NOTE**: the permutaiton step might take some time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "990f6397",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "example/enrichr_kegg_permutation_1000.txt does not exist, initializing permutation\n",
      "Permutation 0\n",
      "Permutation 50\n",
      "Permutation 100\n",
      "Permutation 150\n",
      "Permutation 200\n",
      "Permutation 250\n",
      "Permutation 300\n",
      "Permutation 350\n",
      "Permutation 400\n",
      "Permutation 450\n",
      "Permutation 500\n",
      "Permutation 550\n",
      "Permutation 600\n",
      "Permutation 650\n",
      "Permutation 700\n",
      "Permutation 750\n",
      "Permutation 800\n",
      "Permutation 850\n",
      "Permutation 900\n",
      "Permutation 950\n",
      "Permutation result written to  example/enrichr_kegg_permutation_1000.txt\n",
      "Performing fisher test now...\n",
      "Results written to  example/enrichr_up.txt\n",
      "********************\n",
      "example/enrichr_kegg_permutation_1000.txt already exists, reading file now...\n",
      "Performing fisher test now...\n",
      "Results written to  example/enrichr_down.txt\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "run_enrichr(pathway_dict=pruned_pathway_dict, gene_set=up_regulated_gene_set,\n",
    "            gene_universe=gene_universe, out_file_name='example/enrichr_up.txt', \n",
    "            permutation_num=1000, permutation_file_name='example/enrichr_kegg_permutation_1000.txt')\n",
    "run_enrichr(pathway_dict=pruned_pathway_dict, gene_set=down_regulated_gene_set,\n",
    "            gene_universe=gene_universe, out_file_name='example/enrichr_down.txt', \n",
    "            permutation_num=1000, permutation_file_name='example/enrichr_kegg_permutation_1000.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d81a29d",
   "metadata": {},
   "source": [
    "After getting results from three underlying methods, the last step is to run PET. We'll perform this step for both up and down-regulated genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f54e5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10660fb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9545e78b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
